File: two_pointers/count_zero_request_servers.rb
def count_servers(n, logs, x, queries)
    logs.sort_by! { |l| l[1] }
    
    queries_with_index = queries.each_with_index.map { |q, idx| [q, idx] }.sort
    
    left = right = 0
    m = logs.length
    servers_in_window = Hash.new(0)
    ans = Array.new(queries.length, 0)
    
    queries_with_index.each do |end_time, idx|
        start = end_time - x
        
        while right < m && logs[right][1] <= end_time
            servers_in_window[logs[right][0]] += 1
            right += 1
        end
        
        while left < right && logs[left][1] < start
            servers_in_window[logs[left][0]] -= 1
            servers_in_window.delete(logs[left][0]) if servers_in_window[logs[left][0]] == 0
            left += 1
        end
        
        ans[idx] = n - servers_in_window.size
    end
    
    ans
end

File: two_pointers/count_zero_request_servers.py
class Solution:
    def countServers(self, n: int, logs: List[List[int]], x: int, queries: List[int]) -> List[int]:
        logs.sort(key=lambda l:l[1])
        
        queries = [(q, idx) for idx, q in enumerate(queries)]
        queries.sort()

        
        left, right = 0, 0
        m = len(logs)
        servers_in_window_counts = defaultdict(int)
        ans = [0] * (len(queries))
        for end, idx in queries:
            start = end - x

            while right < m and logs[right][1] <= end:
                servers_in_window_counts[logs[right][0]] += 1
                right += 1

            while left < right and logs[left][1] < start: 
                servers_in_window_counts[logs[left][0]] -= 1
                if servers_in_window_counts[logs[left][0]] == 0:
                    del servers_in_window_counts[logs[left][0]]
                left += 1

            ans[idx] = n - len(servers_in_window_counts)

        return ans